
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Throwable - OpenJDK 21 - W3cubDocs</title>
  
  <meta name="description" content="Instances of two subclasses, Error and Exception, are conventionally used to indicate that exceptional situations have occurred. Typically, these &hellip;">
  <meta name="keywords" content="class, throwable, openjdk, openjdk~21">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/openjdk~21/java.base/java/lang/throwable.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-c8f743fc70bb84079c3cf4b71f5f99835529659c30689b2fd6193651a680f383ef1d96896f8966c7ca8a0160ffd557685bb8ad12820ec6002d60143812238086.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/openjdk~21.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/openjdk~21/" class="_nav-link" title="" style="margin-left:0;">OpenJDK 21</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _openjdk">
				
				
<h1 title="Class Throwable" class="title">Class Throwable</h1> <div class="inheritance" title="Inheritance Tree">
<a href="object" title="class in java.lang">java.lang.Object</a> <div class="inheritance">java.lang.Throwable</div> </div> <section class="class-description" id="class-description"> <dl class="notes"> <dt>All Implemented Interfaces:</dt> <dd><code><a href="../io/serializable" title="interface in java.io">Serializable</a></code></dd> </dl> <dl class="notes"> <dt>Direct Known Subclasses:</dt> <dd>
<code><a href="error" title="class in java.lang">Error</a></code>, <code><a href="exception" title="class in java.lang">Exception</a></code>
</dd> </dl>  <pre class="lang-java" data-language="java">public class Throwable extends Object implements Serializable</pre> <div class="block">The <code>Throwable</code> class is the superclass of all errors and exceptions in the Java language. Only objects that are instances of this class (or one of its subclasses) are thrown by the Java Virtual Machine or can be thrown by the Java <code>throw</code> statement. Similarly, only this class or one of its subclasses can be the argument type in a <code>catch</code> clause. For the purposes of compile-time checking of exceptions, <code>
 Throwable</code> and any subclass of <code>Throwable</code> that is not also a subclass of either <a href="runtimeexception" title="class in java.lang"><code>RuntimeException</code></a> or <a href="error" title="class in java.lang"><code>Error</code></a> are regarded as checked exceptions. <p>Instances of two subclasses, <a href="error" title="class in java.lang"><code>Error</code></a> and <a href="exception" title="class in java.lang"><code>Exception</code></a>, are conventionally used to indicate that exceptional situations have occurred. Typically, these instances are freshly created in the context of the exceptional situation so as to include relevant information (such as stack trace data). </p>
<p>A throwable contains a snapshot of the execution stack of its thread at the time it was created. It can also contain a message string that gives more information about the error. Over time, a throwable can <a href="#addSuppressed(java.lang.Throwable)">suppress</a> other throwables from being propagated. Finally, the throwable can also contain a <i>cause</i>: another throwable that caused this throwable to be constructed. The recording of this causal information is referred to as the <i>chained exception</i> facility, as the cause can, itself, have a cause, and so on, leading to a "chain" of exceptions, each caused by another. </p>
<p>One reason that a throwable may have a cause is that the class that throws it is built atop a lower layered abstraction, and an operation on the upper layer fails due to a failure in the lower layer. It would be bad design to let the throwable thrown by the lower layer propagate outward, as it is generally unrelated to the abstraction provided by the upper layer. Further, doing so would tie the API of the upper layer to the details of its implementation, assuming the lower layer's exception was a checked exception. Throwing a "wrapped exception" (i.e., an exception containing a cause) allows the upper layer to communicate the details of the failure to its caller without incurring either of these shortcomings. It preserves the flexibility to change the implementation of the upper layer without changing its API (in particular, the set of exceptions thrown by its methods). </p>
<p>A second reason that a throwable may have a cause is that the method that throws it must conform to a general-purpose interface that does not permit the method to throw the cause directly. For example, suppose a persistent collection conforms to the <a href="../util/collection" title="interface in java.util"><code>Collection</code></a> interface, and that its persistence is implemented atop <code>java.io</code>. Suppose the internals of the <code>add</code> method can throw an <a href="../io/ioexception" title="class in java.io"><code>IOException</code></a>. The implementation can communicate the details of the <code>IOException</code> to its caller while conforming to the <code>Collection</code> interface by wrapping the <code>IOException</code> in an appropriate unchecked exception. (The specification for the persistent collection should indicate that it is capable of throwing such exceptions.) </p>
<p>A cause can be associated with a throwable in two ways: via a constructor that takes the cause as an argument, or via the <a href="#initCause(java.lang.Throwable)"><code>initCause(Throwable)</code></a> method. New throwable classes that wish to allow causes to be associated with them should provide constructors that take a cause and delegate (perhaps indirectly) to one of the <code>Throwable</code> constructors that takes a cause. Because the <code>initCause</code> method is public, it allows a cause to be associated with any throwable, even a "legacy throwable" whose implementation predates the addition of the exception chaining mechanism to <code>Throwable</code>. </p>
<p>By convention, class <code>Throwable</code> and its subclasses have two constructors, one that takes no arguments and one that takes a <code>String</code> argument that can be used to produce a detail message. Further, those subclasses that might likely have a cause associated with them should have two more constructors, one that takes a <code>Throwable</code> (the cause), and one that takes a <code>String</code> (the detail message) and a <code>Throwable</code> (the cause).</p>
</div> <dl class="notes">
<dt>See <i>Java Language Specification</i>:</dt> <dd> <a href="https://docs.oracle.com/javase/specs/jls/se21/html/jls-11.html#jls-11.2">11.2 Compile-Time Checking of Exceptions</a><br>
</dd> <dt>Since:</dt> <dd>1.0</dd> <dt>See Also:</dt> <dd> <ul class="tag-list"> <li><a href="https://docs.oracle.com/en/java/javase/21/docs/api/serialized-form.html#java.lang.Throwable">Serialized Form</a></li> </ul> </dd> </dl> </section> <section class="summary"> <section class="constructor-summary" id="constructor-summary"> <h2>Constructor Summary</h2>  <table class="summary-table three-column-summary">     <div class="col-constructor-name even-row-color"><code><a href="#%3Cinit%3E()" class="member-name-link">Throwable</a>()</code></div>   <div class="col-constructor-name odd-row-color"><code><a href="#%3Cinit%3E(java.lang.String)" class="member-name-link">Throwable</a><wbr>(<a href="string" title="class in java.lang">String</a>Â message)</wbr></code></div>   <div class="col-constructor-name even-row-color"><code><a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable)" class="member-name-link">Throwable</a><wbr>(<a href="string" title="class in java.lang">String</a>Â message,
 <a href="throwable" title="class in java.lang">Throwable</a>Â cause)</wbr></code></div>   <div class="col-constructor-name odd-row-color"><code><a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable,boolean,boolean)" class="member-name-link">Throwable</a><wbr>(<a href="string" title="class in java.lang">String</a>Â message,
 <a href="throwable" title="class in java.lang">Throwable</a>Â cause,
 booleanÂ enableSuppression,
 booleanÂ writableStackTrace)</wbr></code></div>   <div class="col-constructor-name even-row-color"><code><a href="#%3Cinit%3E(java.lang.Throwable)" class="member-name-link">Throwable</a><wbr>(<a href="throwable" title="class in java.lang">Throwable</a>Â cause)</wbr></code></div>  <tr>
<th>Modifier</th>
<th>Constructor</th>
<th>Description</th>
</tr>
<tr>
<td></td>
<td> <div class="block">Constructs a new throwable with <code>null</code> as its detail message.</div> </td>
</tr>
<tr>
<td></td>
<td> <div class="block">Constructs a new throwable with the specified detail message.</div> </td>
</tr>
<tr>
<td></td>
<td> <div class="block">Constructs a new throwable with the specified detail message and cause.</div> </td>
</tr>
<tr>
<td><code>protected </code></td>
<td> <div class="block">Constructs a new throwable with the specified detail message, cause, <a href="#addSuppressed(java.lang.Throwable)">suppression</a> enabled or disabled, and writable stack trace enabled or disabled.</div> </td>
</tr>
<tr>
<td></td>
<td> <div class="block">Constructs a new throwable with the specified cause and a detail message of <code>(cause==null ? null : cause.toString())</code> (which typically contains the class and detail message of <code>cause</code>).</div> </td>
</tr>
</table> </section><section class="method-summary" id="method-summary"> <h2>Method Summary</h2> <div id="method-summary-table">  <div id="method-summary-table.tabpanel" role="tabpanel"> <table class="summary-table three-column-summary" aria-labelledby="method-summary-table-tab0">                                           <tr>
<th>Modifier and Type</th>
<th>Method</th>
<th>Description</th>
</tr>
<tr>
<td><code>final void</code></td>
<td><code><a href="#addSuppressed(java.lang.Throwable)" class="member-name-link">addSuppressed</a><wbr>(<a href="throwable" title="class in java.lang">Throwable</a>Â exception)</wbr></code></td>
<td> <div class="block">Appends the specified exception to the exceptions that were suppressed in order to deliver this exception.</div> </td>
</tr>
<tr>
<td><code><a href="throwable" title="class in java.lang">Throwable</a></code></td>
<td><code><a href="#fillInStackTrace()" class="member-name-link">fillInStackTrace</a>()</code></td>
<td> <div class="block">Fills in the execution stack trace.</div> </td>
</tr>
<tr>
<td><code><a href="throwable" title="class in java.lang">Throwable</a></code></td>
<td><code><a href="#getCause()" class="member-name-link">getCause</a>()</code></td>
<td> <div class="block">Returns the cause of this throwable or <code>null</code> if the cause is nonexistent or unknown.</div> </td>
</tr>
<tr>
<td><code><a href="string" title="class in java.lang">String</a></code></td>
<td><code><a href="#getLocalizedMessage()" class="member-name-link">getLocalizedMessage</a>()</code></td>
<td> <div class="block">Creates a localized description of this throwable.</div> </td>
</tr>
<tr>
<td><code><a href="string" title="class in java.lang">String</a></code></td>
<td><code><a href="#getMessage()" class="member-name-link">getMessage</a>()</code></td>
<td> <div class="block">Returns the detail message string of this throwable.</div> </td>
</tr>
<tr>
<td><code><a href="stacktraceelement" title="class in java.lang">StackTraceElement</a>[]</code></td>
<td><code><a href="#getStackTrace()" class="member-name-link">getStackTrace</a>()</code></td>
<td> <div class="block">Provides programmatic access to the stack trace information printed by <a href="#printStackTrace()"><code>printStackTrace()</code></a>.</div> </td>
</tr>
<tr>
<td><code>final <a href="throwable" title="class in java.lang">Throwable</a>[]</code></td>
<td><code><a href="#getSuppressed()" class="member-name-link">getSuppressed</a>()</code></td>
<td> <div class="block">Returns an array containing all of the exceptions that were suppressed, typically by the <code>try</code>-with-resources statement, in order to deliver this exception.</div> </td>
</tr>
<tr>
<td><code><a href="throwable" title="class in java.lang">Throwable</a></code></td>
<td><code><a href="#initCause(java.lang.Throwable)" class="member-name-link">initCause</a><wbr>(<a href="throwable" title="class in java.lang">Throwable</a>Â cause)</wbr></code></td>
<td> <div class="block">Initializes the <i>cause</i> of this throwable to the specified value.</div> </td>
</tr>
<tr>
<td><code>void</code></td>
<td><code><a href="#printStackTrace()" class="member-name-link">printStackTrace</a>()</code></td>
<td> <div class="block">Prints this throwable and its backtrace to the standard error stream.</div> </td>
</tr>
<tr>
<td><code>void</code></td>
<td><code><a href="#printStackTrace(java.io.PrintStream)" class="member-name-link">printStackTrace</a><wbr>(<a href="../io/printstream" title="class in java.io">PrintStream</a>Â s)</wbr></code></td>
<td> <div class="block">Prints this throwable and its backtrace to the specified print stream.</div> </td>
</tr>
<tr>
<td><code>void</code></td>
<td><code><a href="#printStackTrace(java.io.PrintWriter)" class="member-name-link">printStackTrace</a><wbr>(<a href="../io/printwriter" title="class in java.io">PrintWriter</a>Â s)</wbr></code></td>
<td> <div class="block">Prints this throwable and its backtrace to the specified print writer.</div> </td>
</tr>
<tr>
<td><code>void</code></td>
<td><code><a href="#setStackTrace(java.lang.StackTraceElement%5B%5D)" class="member-name-link">setStackTrace</a><wbr>(<a href="stacktraceelement" title="class in java.lang">StackTraceElement</a>[]Â stackTrace)</wbr></code></td>
<td> <div class="block">Sets the stack trace elements that will be returned by <a href="#getStackTrace()"><code>getStackTrace()</code></a> and printed by <a href="#printStackTrace()"><code>printStackTrace()</code></a> and related methods.</div> </td>
</tr>
<tr>
<td><code><a href="string" title="class in java.lang">String</a></code></td>
<td><code><a href="#toString()" class="member-name-link">toString</a>()</code></td>
<td> <div class="block">Returns a short description of this throwable.</div> </td>
</tr>
</table> </div> </div> <div class="inherited-list"> <h3 id="methods-inherited-from-class-java.lang.Object">Methods declared in class java.lang.<a href="object" title="class in java.lang">Object</a>
</h3> <code><a href="object#clone()">clone</a>, <a href="object#equals(java.lang.Object)">equals</a>, <a href="object#finalize()">finalize</a>, <a href="object#getClass()">getClass</a>, <a href="object#hashCode()">hashCode</a>, <a href="object#notify()">notify</a>, <a href="object#notifyAll()">notifyAll</a>, <a href="object#wait()">wait</a>, <a href="object#wait(long)">wait</a>, <a href="object#wait(long,int)">wait</a></code>
</div> </section> </section> <section class="details"> <section class="constructor-details" id="constructor-detail"> <h2>Constructor Details</h2>  </section><section class="detail" id="&lt;init&gt;()"> <h3>Throwable</h3> <pre class="lang-java" data-language="java">public Throwable()</pre> <div class="block">Constructs a new throwable with <code>null</code> as its detail message. The cause is not initialized, and may subsequently be initialized by a call to <a href="#initCause(java.lang.Throwable)"><code>initCause(java.lang.Throwable)</code></a>. <p>The <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a> method is called to initialize the stack trace data in the newly created throwable.</p>
</div> </section><section class="detail" id="&lt;init&gt;(java.lang.String)"> <h3>Throwable</h3> <pre class="lang-java" data-language="java">public Throwable(String message)</pre> <div class="block">Constructs a new throwable with the specified detail message. The cause is not initialized, and may subsequently be initialized by a call to <a href="#initCause(java.lang.Throwable)"><code>initCause(java.lang.Throwable)</code></a>. <p>The <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a> method is called to initialize the stack trace data in the newly created throwable.</p>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>message</code> - the detail message. The detail message is saved for later retrieval by the <a href="#getMessage()"><code>getMessage()</code></a> method.</dd> </dl> </section><section class="detail" id="&lt;init&gt;(java.lang.String,java.lang.Throwable)"> <h3>Throwable</h3> <pre class="lang-java" data-language="java">public Throwable(String message, Throwable cause)</pre> <div class="block">Constructs a new throwable with the specified detail message and cause. <p>Note that the detail message associated with <code>cause</code> is <i>not</i> automatically incorporated in this throwable's detail message. </p>
<p>The <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a> method is called to initialize the stack trace data in the newly created throwable.</p>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>message</code> - the detail message (which is saved for later retrieval by the <a href="#getMessage()"><code>getMessage()</code></a> method).</dd> <dd>
<code>cause</code> - the cause (which is saved for later retrieval by the <a href="#getCause()"><code>getCause()</code></a> method). (A <code>null</code> value is permitted, and indicates that the cause is nonexistent or unknown.)</dd> <dt>Since:</dt> <dd>1.4</dd> </dl> </section><section class="detail" id="&lt;init&gt;(java.lang.Throwable)"> <h3>Throwable</h3> <pre class="lang-java" data-language="java">public Throwable(Throwable cause)</pre> <div class="block">Constructs a new throwable with the specified cause and a detail message of <code>(cause==null ? null : cause.toString())</code> (which typically contains the class and detail message of <code>cause</code>). This constructor is useful for throwables that are little more than wrappers for other throwables (for example, <a href="../security/privilegedactionexception" title="class in java.security"><code>PrivilegedActionException</code></a>). <p>The <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a> method is called to initialize the stack trace data in the newly created throwable.</p>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>cause</code> - the cause (which is saved for later retrieval by the <a href="#getCause()"><code>getCause()</code></a> method). (A <code>null</code> value is permitted, and indicates that the cause is nonexistent or unknown.)</dd> <dt>Since:</dt> <dd>1.4</dd> </dl> </section><section class="detail" id="&lt;init&gt;(java.lang.String,java.lang.Throwable,boolean,boolean)"> <h3>Throwable</h3> <pre class="lang-java" data-language="java">protected Throwable(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace)</pre> <div class="block">Constructs a new throwable with the specified detail message, cause, <a href="#addSuppressed(java.lang.Throwable)">suppression</a> enabled or disabled, and writable stack trace enabled or disabled. If suppression is disabled, <a href="#getSuppressed()"><code>getSuppressed()</code></a> for this object will return a zero-length array and calls to <a href="#addSuppressed(java.lang.Throwable)"><code>addSuppressed(java.lang.Throwable)</code></a> that would otherwise append an exception to the suppressed list will have no effect. If the writable stack trace is false, this constructor will not call <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a>, a <code>null</code> will be written to the <code>stackTrace</code> field, and subsequent calls to <code>
 fillInStackTrace</code> and <a href="#setStackTrace(java.lang.StackTraceElement%5B%5D)"><code>setStackTrace(StackTraceElement[])</code></a> will not set the stack trace. If the writable stack trace is false, <a href="#getStackTrace()"><code>getStackTrace()</code></a> will return a zero length array. <p>Note that the other constructors of <code>Throwable</code> treat suppression as being enabled and the stack trace as being writable. Subclasses of <code>Throwable</code> should document any conditions under which suppression is disabled and document conditions under which the stack trace is not writable. Disabling of suppression should only occur in exceptional circumstances where special requirements exist, such as a virtual machine reusing exception objects under low-memory situations. Circumstances where a given exception object is repeatedly caught and rethrown, such as to implement control flow between two sub-systems, is another situation where immutable throwable objects would be appropriate.</p>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>message</code> - the detail message.</dd> <dd>
<code>cause</code> - the cause. (A <code>null</code> value is permitted, and indicates that the cause is nonexistent or unknown.)</dd> <dd>
<code>enableSuppression</code> - whether or not suppression is enabled</dd> <dd>
<code>writableStackTrace</code> - whether or not the stack trace is writable</dd> <dt>Since:</dt> <dd>1.7</dd> <dt>See Also:</dt> <dd> <ul class="tag-list"> <li><a href="outofmemoryerror" title="class in java.lang"><code>OutOfMemoryError</code></a></li> <li><a href="nullpointerexception" title="class in java.lang"><code>NullPointerException</code></a></li> <li><a href="arithmeticexception" title="class in java.lang"><code>ArithmeticException</code></a></li> </ul> </dd> </dl> </section><section class="method-details" id="method-detail"> <h2>Method Details</h2>  </section><section class="detail" id="getMessage()"> <h3>getMessage</h3> <pre class="lang-java" data-language="java">public String getMessage()</pre> <div class="block">Returns the detail message string of this throwable.</div> <dl class="notes"> <dt>Returns:</dt> <dd>the detail message string of this <code>Throwable</code> instance (which may be <code>null</code>).</dd> </dl> </section><section class="detail" id="getLocalizedMessage()"> <h3>getLocalizedMessage</h3> <pre class="lang-java" data-language="java">public String getLocalizedMessage()</pre> <div class="block">Creates a localized description of this throwable. Subclasses may override this method in order to produce a locale-specific message. For subclasses that do not override this method, the default implementation returns the same result as <code>getMessage()</code>.</div> <dl class="notes"> <dt>Returns:</dt> <dd>The localized description of this throwable.</dd> <dt>Since:</dt> <dd>1.1</dd> </dl> </section><section class="detail" id="getCause()"> <h3>getCause</h3> <pre class="lang-java" data-language="java">public Throwable getCause()</pre> <div class="block">Returns the cause of this throwable or <code>null</code> if the cause is nonexistent or unknown. (The cause is the throwable that caused this throwable to get thrown.) <p>This implementation returns the cause that was supplied via one of the constructors requiring a <code>Throwable</code>, or that was set after creation with the <a href="#initCause(java.lang.Throwable)"><code>initCause(Throwable)</code></a> method. While it is typically unnecessary to override this method, a subclass can override it to return a cause set by some other means. This is appropriate for a "legacy chained throwable" that predates the addition of chained exceptions to <code>Throwable</code>. Note that it is <i>not</i> necessary to override any of the <code>PrintStackTrace</code> methods, all of which invoke the <code>getCause</code> method to determine the cause of a throwable.</p>
</div> <dl class="notes"> <dt>Returns:</dt> <dd>the cause of this throwable or <code>null</code> if the cause is nonexistent or unknown.</dd> <dt>Since:</dt> <dd>1.4</dd> </dl> </section><section class="detail" id="initCause(java.lang.Throwable)"> <h3>initCause</h3> <pre class="lang-java" data-language="java">public Throwable initCause(Throwable cause)</pre> <div class="block">Initializes the <i>cause</i> of this throwable to the specified value. (The cause is the throwable that caused this throwable to get thrown.) <p>This method can be called at most once. It is generally called from within the constructor, or immediately after creating the throwable. If this throwable was created with <a href="#%3Cinit%3E(java.lang.Throwable)"><code>Throwable(Throwable)</code></a> or <a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable)"><code>Throwable(String,Throwable)</code></a>, this method cannot be called even once. </p>
<p>An example of using this method on a legacy throwable type without other support for setting the cause is: </p>
<pre>
 try {
     lowLevelOp();
 } catch (LowLevelException le) {
     throw (HighLevelException)
           new HighLevelException().initCause(le); // Legacy constructor
 }
 </pre>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>cause</code> - the cause (which is saved for later retrieval by the <a href="#getCause()"><code>getCause()</code></a> method). (A <code>null</code> value is permitted, and indicates that the cause is nonexistent or unknown.)</dd> <dt>Returns:</dt> <dd>a reference to this <code>Throwable</code> instance.</dd> <dt>Throws:</dt> <dd>
<code><a href="illegalargumentexception" title="class in java.lang">IllegalArgumentException</a></code> - if <code>cause</code> is this throwable. (A throwable cannot be its own cause.)</dd> <dd>
<code><a href="illegalstateexception" title="class in java.lang">IllegalStateException</a></code> - if this throwable was created with <a href="#%3Cinit%3E(java.lang.Throwable)"><code>Throwable(Throwable)</code></a> or <a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable)"><code>Throwable(String,Throwable)</code></a>, or this method has already been called on this throwable.</dd> <dt>Since:</dt> <dd>1.4</dd> </dl> </section><section class="detail" id="toString()"> <h3>toString</h3> <pre class="lang-java" data-language="java">public String toString()</pre> <div class="block">Returns a short description of this throwable. The result is the concatenation of: <ul> <li> the <a href="class#getName()">name</a> of the class of this object </li>
<li> ": " (a colon and a space) </li>
<li> the result of invoking this object's <a href="#getLocalizedMessage()"><code>getLocalizedMessage()</code></a> method </li>
</ul> If <code>getLocalizedMessage</code> returns <code>null</code>, then just the class name is returned.</div> <dl class="notes"> <dt>Overrides:</dt> <dd>
<code><a href="object#toString()">toString</a></code> in class <code><a href="object" title="class in java.lang">Object</a></code>
</dd> <dt>Returns:</dt> <dd>a string representation of this throwable.</dd> </dl> </section><section class="detail" id="printStackTrace()"> <h3>printStackTrace</h3> <pre class="lang-java" data-language="java">public void printStackTrace()</pre> <div class="block">Prints this throwable and its backtrace to the standard error stream. This method prints a stack trace for this <code>Throwable</code> object on the error output stream that is the value of the field <code>System.err</code>. The first line of output contains the result of the <a href="#toString()"><code>toString()</code></a> method for this object. Remaining lines represent data previously recorded by the method <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a>. The format of this information depends on the implementation, but the following example may be regarded as typical: <blockquote><pre>
 java.lang.NullPointerException
         at MyClass.mash(MyClass.java:9)
         at MyClass.crunch(MyClass.java:6)
         at MyClass.main(MyClass.java:3)
 </pre></blockquote> This example was produced by running the program: <pre>
 class MyClass {
     public static void main(String[] args) {
         crunch(null);
     }
     static void crunch(int[] a) {
         mash(a);
     }
     static void mash(int[] b) {
         System.out.println(b[0]);
     }
 }
 </pre> The backtrace for a throwable with an initialized, non-null cause should generally include the backtrace for the cause. The format of this information depends on the implementation, but the following example may be regarded as typical: <pre>
 HighLevelException: MidLevelException: LowLevelException
         at Junk.a(Junk.java:13)
         at Junk.main(Junk.java:4)
 Caused by: MidLevelException: LowLevelException
         at Junk.c(Junk.java:23)
         at Junk.b(Junk.java:17)
         at Junk.a(Junk.java:11)
         ... 1 more
 Caused by: LowLevelException
         at Junk.e(Junk.java:30)
         at Junk.d(Junk.java:27)
         at Junk.c(Junk.java:21)
         ... 3 more
 </pre> Note the presence of lines containing the characters <code>"..."</code>. These lines indicate that the remainder of the stack trace for this exception matches the indicated number of frames from the bottom of the stack trace of the exception that was caused by this exception (the "enclosing" exception). This shorthand can greatly reduce the length of the output in the common case where a wrapped exception is thrown from the same method as the "causative exception" is caught. The above example was produced by running the program: <pre>
 public class Junk {
     public static void main(String args[]) {
         try {
             a();
         } catch(HighLevelException e) {
             e.printStackTrace();
         }
     }
     static void a() throws HighLevelException {
         try {
             b();
         } catch(MidLevelException e) {
             throw new HighLevelException(e);
         }
     }
     static void b() throws MidLevelException {
         c();
     }
     static void c() throws MidLevelException {
         try {
             d();
         } catch(LowLevelException e) {
             throw new MidLevelException(e);
         }
     }
     static void d() throws LowLevelException {
        e();
     }
     static void e() throws LowLevelException {
         throw new LowLevelException();
     }
 }

 class HighLevelException extends Exception {
     HighLevelException(Throwable cause) { super(cause); }
 }

 class MidLevelException extends Exception {
     MidLevelException(Throwable cause)  { super(cause); }
 }

 class LowLevelException extends Exception {
 }
 </pre> As of release 7, the platform supports the notion of <i>suppressed exceptions</i> (in conjunction with the <code>
 try</code>-with-resources statement). Any exceptions that were suppressed in order to deliver an exception are printed out beneath the stack trace. The format of this information depends on the implementation, but the following example may be regarded as typical: <pre>
 Exception in thread "main" java.lang.Exception: Something happened
         at Foo.bar(Foo.java:10)
         at Foo.main(Foo.java:5)
         Suppressed: Resource$CloseFailException: Resource ID = 0
                 at Resource.close(Resource.java:26)
                 at Foo.bar(Foo.java:9)
                 ... 1 more
 </pre> Note that the "... n more" notation is used on suppressed exceptions just as it is used on causes. Unlike causes, suppressed exceptions are indented beyond their "containing exceptions." <p>An exception can have both a cause and one or more suppressed exceptions: </p>
<pre>
 Exception in thread "main" java.lang.Exception: Main block
         at Foo3.main(Foo3.java:7)
         Suppressed: Resource$CloseFailException: Resource ID = 2
                 at Resource.close(Resource.java:26)
                 at Foo3.main(Foo3.java:5)
         Suppressed: Resource$CloseFailException: Resource ID = 1
                 at Resource.close(Resource.java:26)
                 at Foo3.main(Foo3.java:5)
 Caused by: java.lang.Exception: I did it
         at Foo3.main(Foo3.java:8)
 </pre> Likewise, a suppressed exception can have a cause: <pre>
 Exception in thread "main" java.lang.Exception: Main block
         at Foo4.main(Foo4.java:6)
         Suppressed: Resource2$CloseFailException: Resource ID = 1
                 at Resource2.close(Resource2.java:20)
                 at Foo4.main(Foo4.java:5)
         Caused by: java.lang.Exception: Rats, you caught me
                 at Resource2$CloseFailException.&lt;init&gt;(Resource2.java:45)
                 ... 2 more
 </pre>
</div> </section><section class="detail" id="printStackTrace(java.io.PrintStream)"> <h3>printStackTrace</h3> <pre class="lang-java" data-language="java">public void printStackTrace(PrintStream s)</pre> <div class="block">Prints this throwable and its backtrace to the specified print stream.</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>s</code> - <code>PrintStream</code> to use for output</dd> </dl> </section><section class="detail" id="printStackTrace(java.io.PrintWriter)"> <h3>printStackTrace</h3> <pre class="lang-java" data-language="java">public void printStackTrace(PrintWriter s)</pre> <div class="block">Prints this throwable and its backtrace to the specified print writer.</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>s</code> - <code>PrintWriter</code> to use for output</dd> <dt>Since:</dt> <dd>1.1</dd> </dl> </section><section class="detail" id="fillInStackTrace()"> <h3>fillInStackTrace</h3> <pre class="lang-java" data-language="java">public Throwable fillInStackTrace()</pre> <div class="block">Fills in the execution stack trace. This method records within this <code>Throwable</code> object information about the current state of the stack frames for the current thread. <p>If the stack trace of this <code>Throwable</code> <a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable,boolean,boolean)">is not writable</a>, calling this method has no effect.</p>
</div> <dl class="notes"> <dt>Returns:</dt> <dd>a reference to this <code>Throwable</code> instance.</dd> <dt>See Also:</dt> <dd> <ul class="tag-list"> <li><a href="#printStackTrace()"><code>printStackTrace()</code></a></li> </ul> </dd> </dl> </section><section class="detail" id="getStackTrace()"> <h3>getStackTrace</h3> <pre class="lang-java" data-language="java">public StackTraceElement[] getStackTrace()</pre> <div class="block">Provides programmatic access to the stack trace information printed by <a href="#printStackTrace()"><code>printStackTrace()</code></a>. Returns an array of stack trace elements, each representing one stack frame. The zeroth element of the array (assuming the array's length is non-zero) represents the top of the stack, which is the last method invocation in the sequence. Typically, this is the point at which this throwable was created and thrown. The last element of the array (assuming the array's length is non-zero) represents the bottom of the stack, which is the first method invocation in the sequence. <p>Some virtual machines may, under some circumstances, omit one or more stack frames from the stack trace. In the extreme case, a virtual machine that has no stack trace information concerning this throwable is permitted to return a zero-length array from this method. Generally speaking, the array returned by this method will contain one element for every frame that would be printed by <code>printStackTrace</code>. Writes to the returned array do not affect future calls to this method.</p>
</div> <dl class="notes"> <dt>Returns:</dt> <dd>an array of stack trace elements representing the stack trace pertaining to this throwable.</dd> <dt>Since:</dt> <dd>1.4</dd> </dl> </section><section class="detail" id="setStackTrace(java.lang.StackTraceElement[])"> <h3>setStackTrace</h3> <pre class="lang-java" data-language="java">public void setStackTrace(StackTraceElement[] stackTrace)</pre> <div class="block">Sets the stack trace elements that will be returned by <a href="#getStackTrace()"><code>getStackTrace()</code></a> and printed by <a href="#printStackTrace()"><code>printStackTrace()</code></a> and related methods. This method, which is designed for use by RPC frameworks and other advanced systems, allows the client to override the default stack trace that is either generated by <a href="#fillInStackTrace()"><code>fillInStackTrace()</code></a> when a throwable is constructed or deserialized when a throwable is read from a serialization stream. <p>If the stack trace of this <code>Throwable</code> <a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable,boolean,boolean)">is not writable</a>, calling this method has no effect other than validating its argument.</p>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>stackTrace</code> - the stack trace elements to be associated with this <code>Throwable</code>. The specified array is copied by this call; changes in the specified array after the method invocation returns will have no effect on this <code>Throwable</code>'s stack trace.</dd> <dt>Throws:</dt> <dd>
<code><a href="nullpointerexception" title="class in java.lang">NullPointerException</a></code> - if <code>stackTrace</code> is <code>null</code> or if any of the elements of <code>stackTrace</code> are <code>null</code>
</dd> <dt>Since:</dt> <dd>1.4</dd> </dl> </section><section class="detail" id="addSuppressed(java.lang.Throwable)"> <h3>addSuppressed</h3> <pre class="lang-java" data-language="java">public final void addSuppressed(Throwable exception)</pre> <div class="block">Appends the specified exception to the exceptions that were suppressed in order to deliver this exception. This method is thread-safe and typically called (automatically and implicitly) by the <code>try</code>-with-resources statement. <p>The suppression behavior is enabled <em>unless</em> disabled <a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable,boolean,boolean)">via a constructor</a>. When suppression is disabled, this method does nothing other than to validate its argument. </p>
<p>Note that when one exception <a href="#initCause(java.lang.Throwable)">causes</a> another exception, the first exception is usually caught and then the second exception is thrown in response. In other words, there is a causal connection between the two exceptions. In contrast, there are situations where two independent exceptions can be thrown in sibling code blocks, in particular in the <code>try</code> block of a <code>try</code>-with-resources statement and the compiler-generated <code>finally</code> block which closes the resource. In these situations, only one of the thrown exceptions can be propagated. In the <code>try</code>-with-resources statement, when there are two such exceptions, the exception originating from the <code>try</code> block is propagated and the exception from the <code>finally</code> block is added to the list of exceptions suppressed by the exception from the <code>try</code> block. As an exception unwinds the stack, it can accumulate multiple suppressed exceptions. </p>
<p>An exception may have suppressed exceptions while also being caused by another exception. Whether or not an exception has a cause is semantically known at the time of its creation, unlike whether or not an exception will suppress other exceptions which is typically only determined after an exception is thrown. </p>
<p>Note that programmer written code is also able to take advantage of calling this method in situations where there are multiple sibling exceptions and only one can be propagated.</p>
</div> <dl class="notes"> <dt>Parameters:</dt> <dd>
<code>exception</code> - the exception to be added to the list of suppressed exceptions</dd> <dt>Throws:</dt> <dd>
<code><a href="illegalargumentexception" title="class in java.lang">IllegalArgumentException</a></code> - if <code>exception</code> is this throwable; a throwable cannot suppress itself.</dd> <dd>
<code><a href="nullpointerexception" title="class in java.lang">NullPointerException</a></code> - if <code>exception</code> is <code>null</code>
</dd> <dt>Since:</dt> <dd>1.7</dd> </dl> </section><section class="detail" id="getSuppressed()"> <h3>getSuppressed</h3> <pre class="lang-java" data-language="java">public final Throwable[] getSuppressed()</pre> <div class="block">Returns an array containing all of the exceptions that were suppressed, typically by the <code>try</code>-with-resources statement, in order to deliver this exception. If no exceptions were suppressed or <a href="#%3Cinit%3E(java.lang.String,java.lang.Throwable,boolean,boolean)">suppression is disabled</a>, an empty array is returned. This method is thread-safe. Writes to the returned array do not affect future calls to this method.</div> <dl class="notes"> <dt>Returns:</dt> <dd>an array containing all of the exceptions that were suppressed to deliver this exception.</dd> <dt>Since:</dt> <dd>1.7</dd> </dl> </section> </section><div class="_attribution">
  <p class="_attribution-p">
    Â© 1993, 2023, Oracle and/or its affiliates. All rights reserved.<br>Documentation extracted from Debian's OpenJDK Development Kit package.<br>Licensed under the GNU General Public License, version 2, with the Classpath Exception.<br>Various third party code in OpenJDK is licensed under different licenses (see Debian package).<br>Java and OpenJDK are trademarks or registered trademarks of Oracle and/or its affiliates.<br>
    <a href="https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Throwable.html" class="_attribution-link">https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Throwable.html</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
